File: packages/guards/src/output-filter-bypass.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface OutputFilterBypassOptions {
  action: 'block' | 'warn';
}

const DEFAULT_PATTERNS: RegExp[] = [
  /respond\s+in\s+base64/i,
  /encode\s+your\s+response/i,
  /use\s+rot13/i,
  /spell\s+it\s+backwards/i,
  /use\s+pig\s+latin/i,
  /write\s+in\s+code/i,
  /split\s+each\s+character\s+with/i,
  /use\s+(the\s+)?first\s+letter\s+of\s+each\s+word/i,
  /use\s+unicode\s+characters/i,
  /respond\s+as\s+if\s+you\s+were/i,
  /hypothetically\s+speaking/i,
  /output\s+in\s+hex/i,
  /encode\s+(it|the\s+answer|your\s+reply)\s+as/i,
  /write\s+(it\s+)?backwards/i,
];

export function outputFilterBypass(options: OutputFilterBypassOptions): Guard {
  return {
    name: 'output-filter-bypass',
    version: '0.1.0',
    description: 'Detects attempts to bypass output filters',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of DEFAULT_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'output-filter-bypass',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/output_filter_bypass.py
"""Detect output filter bypass attempts."""

import re
import time

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"respond\s+in\s+base64", re.IGNORECASE,
    ),
    re.compile(
        r"encode\s+your\s+response", re.IGNORECASE,
    ),
    re.compile(
        r"use\s+ROT13", re.IGNORECASE,
    ),
    re.compile(
        r"spell\s+it\s+backwards", re.IGNORECASE,
    ),
    re.compile(
        r"use\s+pig\s+latin", re.IGNORECASE,
    ),
    re.compile(
        r"write\s+in\s+code", re.IGNORECASE,
    ),
    re.compile(
        r"split\s+each\s+character", re.IGNORECASE,
    ),
    re.compile(
        r"use\s+(the\s+)?first\s+letter\s+of"
        r"\s+each\s+word",
        re.IGNORECASE,
    ),
    re.compile(
        r"hypothetically\s+speaking", re.IGNORECASE,
    ),
    re.compile(
        r"output\s+(as|in)\s+(hex|binary|morse)",
        re.IGNORECASE,
    ),
    re.compile(
        r"reverse\s+the\s+(output|text|response)",
        re.IGNORECASE,
    ),
]


class _OutputFilterBypass:
    def __init__(
        self,
        *,
        action: str = "block",
    ) -> None:
        self.name = "output-filter-bypass"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="output-filter-bypass",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Output filter bypass attempt detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "matched_patterns": len(matched),
                    "reason": (
                        "Text contains instructions"
                        " requesting encoded or"
                        " obfuscated output to bypass"
                        " safety filters"
                    ),
                }
                if triggered
                else None
            ),
        )


def output_filter_bypass(
    *,
    action: str = "block",
) -> _OutputFilterBypass:
    return _OutputFilterBypass(action=action)
